Snaps settings page

hmalik88 opened the pull request:
Closes: #11681 

Explanation:  Snaps settings page [[FIGMA](https://www.figma.com/file/nBKUQx0btfQiqVf30UA6KD/Flask-MVP?node-id=336%3A6342)] used to interact with the settings of individual snaps. There are two views: a list view under `SnapListTab` and a singular view with `ViewSnap`.  A user should be able to start & disable a snap and also be able to remove it entirely. They should also be able to disconnect a snap from a connected site.

**NOTE:** We do not have a snaps discover page / repository nor do we have snap verification in place (yet), so that part of the design was omitted for now. Some parts of the work were improvised due to the current design not covering all edge cases. Design will be polished in future iterations.

Manual testing steps:  
  - Start storybook
  - Navigate to `SettingsPage` and click on the "snaps" tab
  - Observe the changes
  - Additionally, it is recommended to observe the changes in a flask build: Run `yarn start --build-type flask` to start a dev build

shanejonas committed:
Add snaps settings page

Fix WIP test-data to use object format

Added snap actions to hook up to background

updated pathname key for snaps view page

added boolean to not render subheader in settings for snaps view page, added settings subheader class to snaps view page

updated locale messages, renamed snap prop to currentSnap, added in title & subtitle and placeholder text for when there are no snaps

added icons folder and included a verified icon for flask

updated snaps test data

updated keys per prettier

made authorship pill self enclosing, added box around pill to introduce padding

added scss file for view-snap and added that to settings page scss file

added isSnapViewPage prop and conditionally display subheader and title

updated story to include backRoute and isSnapViewPage props

fixed spacing

added class to subheader and updated description to use correct color

update test data

updated key for permission div

updated view snap styling

updated view snap to have multiple sections encapsulated by box component

updated locale messages with snapAccess key

added translation call around snap_confirm

added subtext for permissions list section

formatted w/prettier

darkwing commented on app/_locales/en/messages.json:
```diff
@@ -2597,6 +2597,9 @@
     "message": "Slow"
   },
+  "snapAccess": {
+    "message": "$1 snap has access to:"
```
Whenever we use a wildcard like `$1`, we add a `description` property to the object which explains what the variable represents.

In this case, I'd make it:

```
"snapAccess": {
  "message": "$1 snap has access to:",
  "description":  "$1 represents the name of the snap"
},
```

darkwing commented on ui/components/app/permissions-connect-permission-list/permissions-connect-permission-list.js:
```diff
@@ -10,15 +10,15 @@ export default function PermissionsConnectPermissionList({ permissions }) {
+      'snap_confirm': {
         leftIcon: 'fas fa-eye',
-        label: 'snap_confirm',
+        label: t('snap_confirm'),
```
Do we have a localization with the key `snap_confirm`?  These are generally camelcased.

rekmarks replied:
They're generally correct, but we will establish keys for all permissioned RPC methods we add, in the vein of `eth_accounts`: https://github.com/MetaMask/metamask-extension/blob/develop/app/_locales/en/messages.json/#L1041-L1044

darkwing commented on ui/components/app/flask/icons/verified-icon.js:
```diff
@@ -0,0 +1,29 @@
+  );
+}
+
+VerifiedIcon.defaultProps = {
```
Instead of using `defaultProps`, I think we can simply set a default value in the function signature:

```
export default function VerifiedIcon({ className, color ='#4CD964' }) {
```

darkwing commented on ui/pages/settings/snaps-list-tab/snap-list-tab.stories.js:
```diff
@@ -0,0 +1,51 @@
+import React, { useState } from 'react';
+import { Provider } from 'react-redux';
+// import { object, boolean, select } from '@storybook/addon-knobs';
```
Do we need this or should we remove?

darkwing commented on ui/pages/settings/snaps-list-tab/view-snap/index.scss:
```diff
@@ -0,0 +1,30 @@
+        display: flex;
+        margin-right: 0px;
+        padding-right: 24px;
+        padding-left: 24px;
```
`padding-inline-start` and `padding-inline-end` would be better for RTL; if there's no padding top or left, we can do `padding: 0 24px;`

darkwing commented on ui/pages/settings/snaps-list-tab/view-snap/view-snap.component.js:
```diff
@@ -0,0 +1,85 @@
+  snap: PropTypes.object,
+  onRemove: PropTypes.func,
+  onToggle: PropTypes.func,
+  // listRoute: PropTypes.string.isRequired,
```
Might need to be removed

darkwing reviewed (changes requested):
Great stuff @hmalik88 !  Just a few comments!

hmalik88 committed:
linter fixes

hmalik88 committed:
made fixes per comments

hmalik88 committed:
moved components into flask directory and updated imports

hmalik88 committed:
updated remove snap section

hmalik88 committed:
added connected sites list

hmalik88 committed:
lint fixes

shanejonas commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,63 @@
+  const envIsPopup = getEnvironmentType() === ENVIRONMENT_TYPE_POPUP;
+  const onClick = (_, snap) => {
+    const route = `${SNAPS_VIEW_ROUTE}/${window.btoa(
+      unescape(encodeURIComponent(snap.name)),
```
i think`snap.name` should be `id`

shanejonas commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,63 @@
+              const decoded = decodeURIComponent(
+                escape(window.atob(pathNameTail)),
+              );
+              return snap.name === decoded;
```
and this would also match for `snap.id`

hmalik88 committed:
style-lint fix

hmalik88 committed:
changed from snap.name -> snap.id

darkwing commented on ui/components/app/connected-sites-list/index.scss:
```diff
@@ -33,6 +33,14 @@
+
+    cursor: pointer;
+    color: $primary-1;
+    padding-left: 24px;
```
`padding-inline-start: 24px` would be a bit better for RTL

darkwing commented on ui/pages/settings/flask/snaps-list-tab/index.scss:
```diff
@@ -0,0 +1,13 @@
+
+.snap-settings-card {
+  margin-top: 8px;
+  margin-bottom: 8px;
```
If there's no margin on the left or right, we can shorthand with `margin: 8px 0`, but no big deal.

shanejonas commented on .storybook/test-data.js:
```diff
@@ -102,6 +102,43 @@ const state = {
     },
+    snaps: {
+      "foo": {
+        name: 'foo',
```
this will have `.id` and `.proposedName`? @rekmarks

rekmarks replied:
Nope! Ref:

```json
{
  "snapStates": {},
  "snaps": {
    "local:http://localhost:8080/": {
      "enabled": true,
      "id": "local:http://localhost:8080/",
      "initialPermissions": {
        "snap_confirm": {}
      },
      "manifest": {
        "description": "An example MetaMask Snap.",
        "initialPermissions": {
          "snap_confirm": {}
        },
        "manifestVersion": "0.1",
        "proposedName": "MetaMask Example Snap",
        "repository": {
          "type": "git",
          "url": "https://github.com/MetaMask/snaps-skunkworks.git"
        },
        "source": {
          "location": {
            "npm": {
              "filePath": "dist/bundle.js",
              "iconPath": "images/icon.svg",
              "packageName": "@metamask/example-snap",
              "registry": "https://registry.npmjs.org/"
            }
          },
          "shasum": "3lEt0yUu080DwV78neROaAAIQWXukSkMnP4OBhOhBnE="
        },
        "version": "0.6.0"
      },
      "permissionName": "wallet_snap_local:http://localhost:8080/",
      "sourceCode": "(...)",
      "status": "stopped",
      "svgIcon": "<svg>...</svg>",
      "version": "0.6.0"
    }
  }
}
```

rekmarks replied:
And `SubjectMetadata`:

```json
{
  "subjectMetadata": {
    "http://localhost:8080": {
      "extensionId": null,
      "iconUrl": null,
      "name": "Hello, Snaps!",
      "origin": "http://localhost:8080",
      "subjectType": "website"
    },
    "https://metamask.github.io": {
      "extensionId": null,
      "iconUrl": null,
      "name": "Snaps Iframe Execution Environment",
      "origin": "https://metamask.github.io",
      "subjectType": "website"
    },
    "local:http://localhost:8080/": {
      "extensionId": null,
      "iconUrl": null,
      "name": "MetaMask Example Snap",
      "origin": "local:http://localhost:8080/",
      "subjectType": "snap",
      "svgIcon": "<svg>...</svg>",
      "version": "0.6.0"
    }
  }
}
```

darkwing commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.container.js:
```diff
@@ -0,0 +1,35 @@
+  const { snap } = ownProps;
+  const subjectMetadata = getSubjectMetadata(state);
+  const connectedSubjects = [];
+  Object.entries(subjectMetadata).forEach(([_, entry]) => {
```
This could probably be a `reduce` instead of `forEach + connectedSubjects` but no big deal, we can improve that in the future.

darkwing commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.container.js:
```diff
@@ -0,0 +1,35 @@
+export default compose(
+  withRouter,
+  connect(mapStateToProps, mapDispatchToProps),
+)(ViewSnap);
```
I think we underlying functional components we avoid the `container` pattern and use methods like `useSelector` and `useDispatch`, but given our rush here, I think this is fine for now.

darkwing reviewed (dismissed):
Lots of nits but none blocking! Well done!

hmalik88 committed:
fixes per comments, updated test data and used appropriate properties for snaps

hmalik88 committed:
updated dispatch call

hmalik88 committed:
removed unnecessary args

rekmarks commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.container.js:
```diff
@@ -0,0 +1,38 @@
+  };
+};
+
+const mapDispatchToProps = (dispatch) => {
+  return {
+    onDisconnect: (origin, address) => {
+      dispatch(removePermittedAccount(origin, address));
+    },
+  };
+};
```
This is using the wrong action. We will need to remove the Snap permission from the origin in question, the user's address has nothing to do with it.

rekmarks replied:
We need to use the `removePermissionsFor` action. Something like this:

```suggestion
const mapDispatchToProps = (dispatch) => {
  return {
    onDisconnect: (connectedOrigin, snapPermissionName) => {
      dispatch(removePermissionsFor({ [connectedOrigin]: [snapPermissionName] }));
    },
  };
};
```

shanejonas commented on app/scripts/metamask-controller.js:
```diff
@@ -1453,6 +1453,9 @@ export default class MetamaskController extends EventEmitter {
       ),
+      disableSnap: this.snapController.disableSnap.bind(this.snapController),
+      enableSnap: this.snapController.enableSnap.bind(this.snapController),
+      removeSnap: this.snapController.removeSnap.bind(this.snapController),
```
```suggestion
      removeSnap: this.snapController.removeSnap.bind(this.snapController),
      revokePermission: this.permissionController.revokePermission.bind(this.permissionController),
```

shanejonas replied:
we need to expose reveokPermission so the ui can call it

shanejonas replied:
or `snapController.removeAllPermissionsFor` (they both do the same thing, removeAllPermissionFor calls revoke underneath)

shanejonas committed:
Fixed some props and action issues with snaps settings page (#13118)

hmalik88 committed:
used idx as key for permission

hmalik88 committed:
added snaps key to messages

hmalik88 committed:
added snaps settings description key

hmalik88 committed:
added code fence

hmalik88 committed:
another code fence

hmalik88 committed:
updated scss, test data, used correct action for onDisconnect

rekmarks reviewed (commented):
@hmalik88, I just started playing around with this to get a feel for it, and the Snap settings UI looks really, really good. Great work!

The one major thing we have to figure out is what to do with Snap settings in the popup UI, because it is currently visually broken. I think it should be reasonably easy to get the main view with the cards into shape, but we may have to redirect to the fullscreen view for the "See Details" page and make that fit into the popup post launch.

rekmarks reviewed (changes requested):
I found another issue, namely that another Snap is showing up in the connected sites list for a Snap.

![image](https://user-images.githubusercontent.com/25517051/146690277-95729d83-cf7a-40d8-ad9e-493f66349941.png)

Looking at the PermissionController state, only a single site (`http://localhost:3000`) has the permission for the Snap I'm looking at in the screenshot, and `npm:@metamask/example-snap` only has the `snap_confirm` permission.

rekmarks reviewed (dismissed):
I also noticed that, when a Snap is removed from settings, the connected sites of the Snap still retain their permissions for the Snap. All permissions for a Snap should be revoked when it is removed, via `PermissionController.revokePermissionForAllSubjects(snapPermissionName)` in the appropriate action / background API function.

hmalik88 committed:
fixes

hmalik88 committed:
removed memoization in snaps settings card, added action to remove snap permissions from all subjects

hmalik88 committed:
updated padding left and right to be 18px

hmalik88 committed:
added tooltip to settings card

hmalik88 committed:
fixed settings routing issue and added another tooltip to the snaps settings card

hmalik88 committed:
removed padding

hmalik88 committed:
removed condition for settings title

hmalik88 committed:
fixed view-snap page

hmalik88 committed:
style lint fixes

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1452,6 +1452,12 @@ export default class MetamaskController extends EventEmitter {
+      disableSnap: this.snapController.disableSnap.bind(this.snapController),
+      enableSnap: this.snapController.enableSnap.bind(this.snapController),
+      removeSnap: this.snapController.removeSnap.bind(this.snapController),
+      revokePermissionsForAllSubjects: this.permissionController.revokePermissionForAllSubjects.bind(
```
Nit: It might make sense to create a `removeSnap` function here in `metamask-controller.js`, and call `revokePermissionsForAllSubjects` immediately after `removeSnap` is called to remove those snap permissions.

Right now the _only_ time that we want to call this function is after calling a snap. And we _never_ want to remove a snap without removing permissions associated with that snap. Moving these two steps together simplifies the background API and makes it less likely we will accidentally remove a snap without removing the associated permissions.

Gudahtt commented on ui/components/app/connected-sites-list/connected-sites-list.component.js:
```diff
@@ -39,11 +40,20 @@ export default class ConnectedSitesList extends Component {
-              title={t('disconnect')}
-              onClick={() => onDisconnect(subject.origin)}
-            />
+            {isSnap ? (
```
Rather than make the "Disconnect" button look different for snaps v.s. the connected sites page, can we instead change this to display the full word "Disconnect" in both places? Or use the trash can for now and worry about this later? It seems silly for this component to be passed an `isSnap` prop for this sole purpose. Nothing about snaps makes "Disconnect" preferable to a trash can icon.

Gudahtt replied:
Also, this is a pre-existing mistake that we can fix in a separate PR, but we should use clickable elements for clickable things. Anchors and buttons are built-in elements that the browser already understands are meant to be interacted with, so they automatically get focused by keyboards and interpreted as interactive by screen readers and other assistive technologies. Patching `onClick` handlers onto divs makes accessibility unnecessarily complicated.

Gudahtt commented on ui/components/app/flask/icons/verified-icon.js:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export default function VerifiedIcon({ className, color = '#4CD964' }) {
```
I guess this is unrelated to the rest of the PR? We don't use this icon anywhere yet. Which is good, since we don't have any verified snaps yet.

hmalik88 replied:
Yes, I saw it part of the design and decided to just make the icon now, since it technically will be a part of the settings page at some point.

Gudahtt commented on ui/components/app/flask/icons/verified-icon.js:
```diff
@@ -0,0 +1,25 @@
+}
+
+VerifiedIcon.propTypes = {
+  className: PropTypes.string,
```
Nit: Including JSDoc comments here for each prop type will improve the docs shown in Storybook for this component.

It's a good practice for all components really, but especially for those in Storybook.

hmalik88 replied:
Will do, overlooked that -- thanks!

Gudahtt commented on ui/components/app/permissions-connect-permission-list/permissions-connect-permission-list.js:
```diff
@@ -113,7 +113,7 @@ export default function PermissionsConnectPermissionList({ permissions }) {
 
         return (
-          <div className="permission" key={label}>
+          <div className="permission" key={idx}>
```
This seems strictly worse. Why use the index as the key? The key should uniquely identify the entry, so that React can tell the difference between permissions getting re-arranged v.s. their properties being changed. The label was already a unique representation of the permission (or so I thought). The permission name would work as well. But the index does not serve as an adequate key, it is not unique per permission.

hmalik88 replied:
The reason I changed this to idx was because you can have multiple permissions returning as `UNKOWN_PERMISSION`, using that label as a key doesn't work. I ran into that problem when the key was as label before.

Gudahtt replied:
Ahhh I see. Well, using an index silences the warning, but it doesn't fix the problem. The index is what React already falls back to internally.

Could we use the permission name instead maybe? That shouldn't be duplicated.

Also, which permissions were coming up as unknown? :thinking: They are supposed to all be listed here.

hmalik88 replied:
I think it may have been part of the test-data, hmmm maybe we don't run into that problem in production since we're cognizant of all the permissions we allow...I'll change it to just `permission` then. Thanks!

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/view-snap/index.scss:
```diff
@@ -0,0 +1,84 @@
+.view-snap-section {
+  .connected-sites-list__content-row {
+    border-top: none;
+    border-bottom: 1px solid #d6d9dc;
```
There is a variable for this color:
```suggestion
    border-bottom: 1px solid $ui-2;
```

hmalik88 replied:
Will change, thanks!

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/view-snap/index.scss:
```diff
@@ -0,0 +1,84 @@
+}
+
+.settings-page__content-item.view-snap-section {
+  border-bottom: 1px solid #d8d8d8;
```
Nit: We should really avoid hard-coding colors into stylesheets. This doesn't appear to be one of the design system colors either. Can we use another color here instead from our set of design system colors? (in `ui/css/design-system/colors.scss`)

hmalik88 replied:
Yeah, no problem -- I'll find an alternative.

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/index.scss:
```diff
@@ -0,0 +1,28 @@
+.snaps-list-wrapper {
```
Nit: These class names do not comply with the naming convention this project uses, which is BEM (Block, Element, Modifier). See here for more information: http://getbem.com/

Typically the component is a "block", and each piece inside the component is an "element". So we would have a top-level class for the block, and sub classes within that block called `&__[element-name]`.

hmalik88 replied:
Ah! I'll make those changes.

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,64 @@
+    currentSnap:
+      state.metamask.snaps && viewingSnap
+        ? Object.entries(state.metamask.snaps)
+            .map(([_, value]) => value)
```
Nit: `value` is not a great variable name. `snap` or `snapState` might be better here.

hmalik88 replied:
I agree.

Gudahtt commented on ui/selectors/permissions.js:
```diff
@@ -96,6 +97,27 @@ export function getConnectedSubjectsForSelectedAddress(state) {
+
+  const connectedSubjects = [];
+
+  Object.entries(subjects).forEach(([subjectKey, { permissions }]) => {
```
Nit: may as well call this parameter `origin`. That is what it is, and would save reassigning it later on line 112.

Gudahtt commented on ui/pages/settings/settings.component.js:
```diff
@@ -152,7 +165,7 @@ class SettingsPage extends PureComponent {
       subheaderText = t(initialBreadCrumbKey);
     } else {
-      subheaderText = t(pathnameI18nKey || 'contacts');
+      subheaderText = t(pathnameI18nKey || 'general');
```
What motivated this change?

hmalik88 replied:
In the settings page container, there isn't a mapping for `SETTINGS_ROUTE`, I think whomever wrote this code was trying to default the settings route to a certain key -- "contacts" is the wrong key here. If you look in expanded view, the subheaderText shows as "Contacts" for when you first navigate to settings, it should be "General".

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,64 @@
+  const viewingSnap = Boolean(pathname.match(SNAPS_VIEW_ROUTE));
+  const pathNameTail = pathname.match(/[^/]+$/u)[0];
+  const envIsPopup = getEnvironmentType() === ENVIRONMENT_TYPE_POPUP;
+  const onClick = (_, snap) => {
```
Well this is highly unconventional :sweat_smile: 

As @darkwing mentioned, this container/component pattern is deprecated and should not be used going forward. But in that pattern, the `mapStateToProps` function was for mapping state to props, as the title implies. These two functions aren't doing that - they don't use state. They can be embedded directly in the component, there is no reason to construct them here.

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.component.js:
```diff
@@ -0,0 +1,136 @@
+            <Box className="view-snap-pill-container" paddingLeft={2}>
+              <SnapsAuthorshipPill
+                packageName={snap.manifest.proposedName}
+                url={snap.manifest.repository.url}
```
The `repository` field is technically optional. See the spec: https://github.com/MetaMask/specifications/blob/main/snaps/publishing.md
```suggestion
                url={snap.manifest.repository?.url}
```

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,64 @@
+  };
+};
+const mapDispatchToProps = (dispatch) => ({
+  onToggle: (_, snap) => {
```
Nit: rather than ignore the first parameter, why not remove it completely? It doesn't need to be there.

Gudahtt replied:
This applies to the `onClick` function defined in `mapStateToProps` as well

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.container.js:
```diff
@@ -0,0 +1,64 @@
+  const envIsPopup = getEnvironmentType() === ENVIRONMENT_TYPE_POPUP;
+  const onClick = (_, snap) => {
+    const route = `${SNAPS_VIEW_ROUTE}/${window.btoa(
+      unescape(encodeURIComponent(snap.id)),
```
Huh. What is going on here? I have never seen `unescape` used before, and apparently it's deprecated.

Why can't we just use `encodeURIComponent(snap.id)` here?

hmalik88 replied:
Hmm not sure, @shanejonas -- is there a specific reason for the use of unescape?

shanejonas replied:
needs escape to handle all cases properly: https://developer.mozilla.org/en-US/docs/Glossary/Base64#the_unicode_problem

Gudahtt replied:
Yeah but why base64 encode it in the first place? `encodeURIComponent` is all we need to make it URL-safe.

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.component.js:
```diff
@@ -0,0 +1,110 @@
+                <SnapSettingsCard
+                  className="snap-settings-card"
+                  isEnabled={snap.enabled}
+                  dateAdded={new Date().toDateString()}
```
This always shows the current date. That's not right. Also this format isn't ideal because it doesn't match the users current locale.

Better to remove the date if we don't know when the snap was added. If we to do know, we should format the date with the `Intl.DateTimeFormat` API, which respects the user's current locale.

Gudahtt replied:
It looks like there is no install date stored in the snaps controller right now, so we can omit this for now.

georgewrmarshall commented on ui/pages/settings/flask/snaps-list-tab/view-snap/index.scss:
```diff
@@ -0,0 +1,84 @@
+
+.settings-page__content-item.view-snap-section {
+  border-bottom: 1px solid #d8d8d8;
+  padding-bottom: 18px;
```
Should adhere to 8px grid unless specific padding is absolutely necessary for layout.
```suggestion
  padding-bottom: 16px;
```

georgewrmarshall reviewed (commented):
UI Looks good! Have a request about the Icon. It doesn't look like it's being used yet? Ping me after you've made @Gudahtt's suggested changes and I will approve

hmalik88 committed:
fixes per comments

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/snap-list-tab.component.js:
```diff
@@ -51,9 +51,9 @@ const SnapListTab = ({
   }
   return (
-    <>
+    <div className="snap-list-tab">
```
Nit: This was fine beforehand - the `snap-list-tab` class doesn't have any rules in it, so it doesn't need to be set anywhere.

hmalik88 committed:
updated actions, moved view snap out of snap-list-tab, updated containers

hmalik88 committed:
style lint fixes

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.component.js:
```diff
@@ -0,0 +1,142 @@
+              css={{
+                maxWidth: '175px',
+              }}
+              onClick={() => {
+                dispatch(removeSnap(snap));
+                history.push(SNAPS_LIST_ROUTE);
+              }}
```
Two issues here: first `removeSnap` is async, and we probably want to wait on that before redirecting so that the user doesn't still see their snap listed. But more importantly, `removeSnap` is not an action creator, so it should not be passed to `dispatch`. It's just a regular function.

```suggestion
              onClick={async () => {
                await removeSnap(snap);
                history.push(SNAPS_LIST_ROUTE);
              }}
```

Gudahtt replied:
Actually... after trying this out, there was still one render of the old removed snap being listed, and it's my fault :facepalm:. The `forceUpdateMetamaskState` call would prevent this, but since that was removed, the state update doesn't propagate to the UI for a few milliseconds, leaving us 1 render cycle with stale data. Oops.

Not a big deal at all, as a user it's so quick it's not confusing in the slightest. Just not ideal.

Gudahtt commented on ui/pages/settings/flask/snaps-list-tab/view-snap/view-snap.container.js:
```diff
@@ -0,0 +1,58 @@
+  return {
+    onDisconnect: (connectedOrigin, snapPermissionName) => {
+      dispatch(
+        removePermissionsFor({
```
Lots of other examples here of calling `dispatch` on things that aren't action creators. `dispatch` isn't needed in any of these cases. These can be called directly from the component actually, rather than passed as props via `mapDispatchToProps`.

hmalik88 replied:
I fixed the others, namely, `enableSnap` and `disableSnap`. Wasn't sure on changing `removePermissionsFor`, doesn't that use `dispatch`?

hmalik88 replied:
On second thought, I just decided to play it safe and revert the changes taking out `forceMetamaskUpdateState`.

hmalik88 committed:
final fixes

hmalik88 committed:
last fixes

Gudahtt committed:
Remove stray console.log

Gudahtt committed:
Update actions to use convenience function

Gudahtt committed:
Pass `permissionName` to `getSubjectsWithPermission` selector

Gudahtt committed:
Make `removeSnap` a synchronous function

Action creators must be synchronous. The function they return can be
async though.

Gudahtt committed:
Move navigation after removing snap to `useEffect` hook

The UI would crash after removing a snap because the snap metadata
object for the current page was no longer present. Instead a
`useEffect` hook is used to navigate away from the page when the snap
is removed.

This has the added benefit of ensuring the page doesn't crash when the
snap is removed in another tab as well.

Gudahtt committed:
Fix lint error accidentally introduced when rebasing

hmalik88 reviewed (commented):
Reviewed @Gudahtt 's commits, looking good! Things are in a working state.

Gudahtt reviewed (approved):
LGTM!
